EpikCord/client/http_client.py

```diff
@@ -1,312 +1,233 @@
-
-            with contextlib.suppress(KeyError):
-                del self.buckets[bucket_hash]
-
-        bucket.close_task.cancel()
-
-        bucket.close_task = asyncio.create_task(dispose())
-
-        return res
-
-    @staticmethod
-    async def log_request(res, body: Optional[dict] = None):
-        message = [
-            f"Sent a {res.request_info.method} to {res.url} "
-            f"and got a {res.status} response. ",
-            f"Content-Type: {res.headers['Content-Type']} ",
-        ]
-
-        if body:
-            message.append(f"Sent body: {body} ")
-
-        if h := dict(res.request_info.headers):
-            message.append(f"Sent headers: {h} ")
-
-        if h := dict(res.headers):
-            message.append(f"Received headers: {h} ")
-
-        try:
-            message.append(f"Received body: {await res.json()} ")
-
-        finally:
-            logger.debug("".join(message))
-
-    def base(
-        self,
-        method,
-        url,
-        *args,
-        to_discord: bool = True,
-        **kwargs,
-    ):
-        if to_discord:
-            return self.request(method, url, *args, **kwargs)
-        return self.session.request(method, url, *args, **kwargs)
-
-    get = partialmethod(base, "GET")
-    post = partialmethod(base, "POST")
-    put = partialmethod(base, "PUT")
-    patch = partialmethod(base, "PATCH")
-    delete = partialmethod(base, "DELETE")
-
-    async def get_gateway(self) -> discord_typings.GetGatewayData:
-        """
-        Get the gateway url.
-        Returns:
-            :class:`GetGatewayData`
-        """
-        res = await self.get("/gateway")
-        return await res.json()
-
-    async def get_gateway_bot(self) -> discord_typings.GetGatewayBotData:
-        """
-        Get information about the bot logging in, including the gateway url.
-        This is useful when sharding.
-
-        Returns:
-            :class:`GetGatewayBotData`
-        """
-        res = await self.get("/gateway/bot")
-        return await res.json()
-
-
-__all__ = ("HTTPClient",)
+                    bucket = Bucket(bucket_hash=response.headers["X-RateLimit-Bucket"])
+                    if bucket in self.buckets.values():
+                        self.buckets[f"{method}:{url}"] = list(self.buckets.values())[
+                            list(self.buckets.values()).index(bucket)
+                        ]
+                        bucket = self.buckets[f"{method}:{url}"]
+                    else:
+                        self.buckets[f"{method}:{url}"] = bucket
+
+            if response.status == 429:
+                data = await response.json()
+                if data["global"]:
+                    self.global_event.clear()
+                bucket.clear()
+                await asyncio.sleep(data["retry_after"])
+                self.global_event.set()
+                bucket.set()
+                return await self.request(method, url, *args, **kwargs)
```

Comment:
Why not use a for loop instead? You can only try 5 times after all. Plus this will flood the call stack.

Reply:
Good idea. I'll do that soon.